### PR TITLE
Add ability to use path wildcard segments

### PR DESCRIPTION
**Pull request by jefferai**
Adds the ability to do path statements like:

`path "foo/+/zip/+/foobarzipzap" {`

### CODE REVIEW

File: vault/acl.go [hunk 1]
```diff
@@ -334,13 +375,109 @@ func (a *ACL) AllowOperation(ctx context.Context, req *logical.Request, capCheck
...
+		permissions = raw.(*ACLPermissions)
+		capabilities = permissions.CapabilitiesBitmap
+		goto CHECK
 	}
-	permissions = raw.(*ACLPermissions)
-	capabilities = permissions.CapabilitiesBitmap
+
+	if len(a.segmentWildcardPaths) > 0 {
+		pathParts := strings.Split(path, "/")
+		for k := range a.segmentWildcardPaths {
```
Comment (chrishoffman) [inline vault/acl.go]: Can we rename k?  I found this hard to read through remembering if k was the ACL path or the request path.

Comment (chrishoffman) [reply vault/acl.go]: And splitK below.

Comment (jefferai) [reply vault/acl.go]: Will do.

**Comment by jefferai** [comment]
Apologies, I'd accidentally pushed that exact change to the wrong branch earlier. I've pushed it here now.

I'll address the splitk thing.

CLA: there doesn't seem to be a way to reset it and it picked up things when I merged in master-oss that were completely irrelevant. I've deleted it for now; the only changes here are from me.

File: vault/acl.go [hunk 2]
```diff
@@ -334,13 +367,81 @@ func (a *ACL) AllowOperation(ctx context.Context, req *logical.Request, capCheck
...
+				}
+				if aclPart != pathParts[i] {
+					// Mismatch, exit out
+					invalid = true
+					break
+				}
+			}
+			// If invalid isn't set then we got through the full segmented path
+			// without finding a mismatch, so it's valid
+			if !invalid {
```
Comment (briankassouf) [inline vault/acl.go]: Might be safer to inverse this and set-if-valid. That way it's more explicit if any more conditions are added above

Comment (jefferai) [reply vault/acl.go]: The reason this is tough to do the other way is that exiting the for loop naturally (by running through the full range) is actually a success condition. If you finish the for loop you don't (easily) know if it's by breaking out or finishing it naturally.

**Pull request merged**